src/index/hnsw/faiss_hnsw.cc

```diff
@@ -0,0 +1,1398 @@
+
+        auto ids = std::make_unique<faiss::idx_t[]>(rows * k);
+        auto distances = std::make_unique<float[]>(rows * k);
+        try {
+            std::vector<folly::Future<folly::Unit>> futs;
+            futs.reserve(rows);
+            for (int64_t i = 0; i < rows; ++i) {
+                futs.emplace_back(search_pool->push([&, idx = i] {
+                    // 1 thread per element
+                    ThreadPool::ScopedOmpSetter setter(1);
+
+                    // set up a query
+                    // const float* cur_query = (const float*)data + idx * dim;
+
+                    const float* cur_query = nullptr;
+
+                    std::vector<float> cur_query_tmp(dim);
+                    if (data_format == DataFormatEnum::fp32) {
+                        cur_query = (const float*)data + idx * dim;
+                    } else {
+                        convert_rows(data, cur_query_tmp.data(), data_format, idx, 1, dim);
+                        cur_query = cur_query_tmp.data();
+                    }
+
+                    // set up local results
+                    faiss::idx_t* const __restrict local_ids = ids.get() + k * idx;
+                    float* const __restrict local_distances = distances.get() + k * idx;
+
+                    // set up faiss search parameters
+                    knowhere::SearchParametersHNSWWrapper hnsw_search_params;
+                    if (hnsw_cfg.ef.has_value()) {
+                        hnsw_search_params.efSearch = hnsw_cfg.ef.value();
+                    }
+                    // do not collect HNSW stats
+                    hnsw_search_params.hnsw_stats = nullptr;
+                    // set up feder
+                    hnsw_search_params.feder = feder_result.get();
+                    // set up kAlpha
+                    hnsw_search_params.kAlpha = bitset.filter_ratio() * 0.7f;
+
+                    // set up a selector
+                    BitsetViewIDSelector bw_idselector(bitset);
+                    faiss::IDSelector* id_selector = (bitset.empty()) ? nullptr : &bw_idselector;
+
+                    hnsw_search_params.sel = id_selector;
+
+                    // use knowhere-based search by default
+                    const bool override_faiss_search = hnsw_cfg.override_faiss_search.value_or(true);
+
+                    // check if we have a refine available.
+                    faiss::IndexRefine* const index_refine = dynamic_cast<faiss::IndexRefine*>(index.get());
+
+                    if (index_refine != nullptr) {
+                        // yes, it is possible to refine results.
+
+                        // cast a base index to IndexHNSW-based index
+                        faiss::IndexHNSW* const index_hnsw = dynamic_cast<faiss::IndexHNSW*>(index_refine->base_index);
+
+                        if (index_hnsw == nullptr) {
+                            // this is unexpected
+                            throw std::runtime_error("Expecting faiss::IndexHNSW");
```

Comment:
We cannot, because the exception is thrown inside one of folly's threads. 
Realistically speaking, this is a safe-guard error, similar to assert, which should never get triggered under normal conditions.

Reply:
IC, should we take care of the folly thread failures?
What I usually do is avoiding throw exception in a folly thread(since the thread will get killed), and set a status that can be verified out side of the folly thread?

Reply:
this is exactly the situation there a thread needs to be killed and report an exception.

Reply:
From the code, this error will kill the folly thread, and exception will throw from `WaitAllSuccess`. Then the catch block outside will catch this error and return an error code to Milvus. User will get error report, do some modification and then retry. But the folly thread is permenentally dead.

1. If this is an unexpected error, we need to crash the Milvus
2. If it is a fixable problem (mis operation order, retryable), then we should not kill the thread.

Reply:
Created a issue https://github.com/zilliztech/knowhere/issues/777 to track this and unblock this PR